processor/src/trace/virtual_bus/circuit/mod.rs

```diff
@@ -0,0 +1,290 @@
+// ================================================================================================
+
+/// Represents a fraction `numerator / denominator` as a pair `(numerator, denominator)`. This is
+/// the type for the gates' inputs in [`prover::EvaluatedCircuit`].
+///
+/// Hence, addition is defined in the natural way fractions are added together: `a/b + c/d = (ad +
+/// bc) / bd`.
+#[derive(Debug, Clone, Copy)]
+pub struct CircuitWire<E: FieldElement> {
+    numerator: E,
+    denominator: E,
+}
+
+impl<E> CircuitWire<E>
+where
+    E: FieldElement,
+{
+    /// Creates new projective coordinates from a numerator and a denominator.
+    pub fn new(numerator: E, denominator: E) -> Self {
+        assert_ne!(denominator, E::ZERO);
+
+        Self {
+            numerator,
+            denominator,
+        }
+    }
+}
+
+impl<E> Add for CircuitWire<E>
+where
+    E: FieldElement,
+{
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let numerator = self.numerator * other.denominator + other.numerator * self.denominator;
+        let denominator = self.denominator * other.denominator;
+
+        Self::new(numerator, denominator)
+    }
+}
+
+/// Converts a main trace row (or more generally "query") to numerators and denominators of the
+/// input layer.
+fn evaluate_fractions_at_main_trace_query<E>(
+    query: &[E],
+    log_up_randomness: &[E],
+) -> [[E; NUM_WIRES_PER_TRACE_ROW]; 2]
+where
+    E: FieldElement,
+{
+    // numerators
+    let multiplicity = query[M_COL_IDX];
+    let f_m = {
+        let mem_selec0 = query[CHIPLETS_OFFSET];
+        let mem_selec1 = query[CHIPLETS_OFFSET + 1];
+        let mem_selec2 = query[CHIPLETS_OFFSET + 2];
+        mem_selec0 * mem_selec1 * (E::ONE - mem_selec2)
+    };
+
+    let f_rc = {
+        let op_bit_4 = query[DECODER_OP_BITS_OFFSET + 4];
+        let op_bit_5 = query[DECODER_OP_BITS_OFFSET + 5];
+        let op_bit_6 = query[DECODER_OP_BITS_OFFSET + 6];
+
+        (E::ONE - op_bit_4) * (E::ONE - op_bit_5) * op_bit_6
+    };
+
+    // denominators
+    let alphas = log_up_randomness;
+
+    let table_denom = alphas[0] - query[V_COL_IDX];
+    let memory_denom_0 = -(alphas[0] - query[MEMORY_D0_COL_IDX]);
+    let memory_denom_1 = -(alphas[0] - query[MEMORY_D1_COL_IDX]);
+    let stack_value_denom_0 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET]);
+    let stack_value_denom_1 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 1]);
+    let stack_value_denom_2 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 2]);
+    let stack_value_denom_3 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 3]);
+
+    [
+        [multiplicity, f_m, f_m, f_rc, f_rc, f_rc, f_rc, E::ZERO],
+        [
+            table_denom,
+            memory_denom_0,
+            memory_denom_1,
+            stack_value_denom_0,
+            stack_value_denom_1,
+            stack_value_denom_2,
+            stack_value_denom_3,
+            E::ONE,
+        ],
+    ]
+}
```

Comment:
Is this basically the only place where we have logic specific to Miden VM? Not for this PR, but I'm wondering if everything else is generic logic that could be a part of Winterfell rather than Miden VM. For example, we could introduce a method on the `Air` trait that would basically do the job of this function and look something like this:
```rust
fn build_gkr_input_layer<F, E>(&self, row: &[F], logup_rands: &[E) -> [Vec<E>; 2]
where
  F: FieldElement<BaseField = Self::BaseField>,
  E: FieldElement<BaseField = Self::BaseField> + ExtensionOf<F>,
```
And then both GKR prover and verifier would be able to use this to build and verify the proof respectively.

A couple of other questions:
- What is `log_up_randomness` here? Is it a set of random values generated from the main trace commitment root? Are these the "lambdas" referenced in [this issue](https://github.com/facebook/winterfell/issues/286) or is it something else?
  - I'm assuming we are using only the first element from `log_up_randomness` because of the nature of the constraints. If we had more complicated tables (e.g., where multiple columns needed to be merged into a single value per row), we'd need to use more random elements - right?
- Where would we specify "boundary" constraints for the buses? Specifically, for sub-busses which do not start and/or end with 0, what would be the way to specify initial/final values?

Reply:
Yes, that makes sense. This will require however that Winterfell is able to handle the logic for building the `s`-column similar to how it does for the `l`-column (i.e., Lagrange column).

To this end we would need to introduce the concept of an oracle. This is basically a wrapper around `CompositionPolynomial` with information about the `index` of the columns appearing in the algebraic expression encoded in the `CompositionPolynomial` as well as additional information needed to verify the final evaluation claim e.g., whether a column is periodic (or more generally transparent) and thus can be computed by the verifier un-assisted and hence does not need to be included in the `s`-column, whether a column is a shift as this is important when building the `s`-column...



> * What is `log_up_randomness` here? Is it a set of random values generated from the main trace commitment root?

Yes, these are the randomness generated from the main trace commitment, usually called the $\alpha$'s and indeed if we had more column to merge we would use more than one.

> Where would we specify "boundary" constraints for the buses? Specifically, for sub-busses which do not start and/or end with 0, what would be the way to specify initial/final values?

The verifier will compute all of the boundary values in order to get a single claim value. The verifier will then check that the output of the LogUp circuit equals that claim. More precisely this would be done here 
```rust
pub fn verify<
    E: FieldElement<BaseField = Felt>,
    C: RandomCoin<Hasher = H, BaseField = Felt>,
    H: ElementHasher<BaseField = Felt>,
>(
    claim: E,
    proof: GkrCircuitProof<E>,
    log_up_randomness: Vec<E>,
    transcript: &mut C,
) -> Result<FinalOpeningClaim<E>, VerifierError> {
...
// check that the output matches the expected `claim`
    if (p0 * q1 + p1 * q0) / (q0 * q1) != claim {
        return Err(VerifierError::MismatchingCircuitOutput);
    }
...
}
```
At the moment, `claim` is set to zero but if we want to go with the proposal of delegating everything to Winterfell, then we would need to add a method to the GKR verifier so that it can compute `claim` from the public inputs.